File: Algorithm/LeetCode/weekly/414_0908/3281_mid.py
from bisect import bisect_left, bisect_right, insort_left, insort_right, insort, bisect
from math import ceil, floor, pow, gcd, sqrt, log10, fabs, fmod, factorial, inf, pi, e
from heapq import heapify, heapreplace, heappush, heappop, heappushpop, nlargest, nsmallest
from collections import defaultdict, Counter, deque
from itertools import permutations, combinations, combinations_with_replacement, accumulate, count, groupby, pairwise
from queue import PriorityQueue, Queue, LifoQueue
from functools import lru_cache, cache, reduce
from typing import List, Optional
import sys

from sortedcontainers import SortedList, SortedDict, SortedSet

sys.setrecursionlimit(10001000)

MOD = int(1e9 + 7)
INF = int(1e20)
INFMIN = float('-inf')  # 负无穷
INFMAX = float('inf')  # 正无穷
PI = 3.141592653589793
direc = [(1, 0), (0, 1), (-1, 0), (0, -1)]
direc8 = [(1, 0), (0, 1), (-1, 0), (0, -1), (1, 1), (1, -1), (-1, 1), (-1, -1)]
ALPS = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
alps = 'abcdefghijklmnopqrstuvwxyz'

# @Author  : https://github.com/hakusai22
# @Time    : 2024/09/08 16:00
# @题目     :
# @参考     :   https://leetcode.cn/problems/maximize-score-of-numbers-in-ranges/solutions/2908931/er-fen-da-an-zui-da-hua-zui-xiao-zhi-pyt-twe2/
# 时间复杂度 :

class Solution:
    def maxPossibleScore(self, start: List[int], d: int) -> int:
        start.sort()

        def check(score: int) -> bool:
            pre_x = -inf  # 也可以只用一个 x 变量
            for s in start:
                x = pre_x + score
                if x > s + d:
                    return False
                pre_x = max(x, s)
            return True

        left, right = 0, (start[-1] + d - start[0]) // (len(start) - 1) + 1
        while left + 1 < right:
            mid = (left + right) // 2
            if check(mid):
                left = mid
            else:
                right = mid
        return left

File: Algorithm/LeetCode/weekly/414_0908/3281_mid.ts
/**
 * @Time 2024/09/08 15:38
 * @Author  : https://github.com/hakusai22
 * 给你一个整数数组 start 和一个整数 d，代表 n 个区间 [start[i], start[i] + d]。
 * 你需要选择 n 个整数，其中第 i 个整数必须属于第 i 个区间。所选整数的 得分 定义为所选整数两两之间的 最小 绝对差。
 * 返回所选整数的 最大可能得分 。
 *
 * 2 <= start.length <= 10^5
 * 0 <= start[i] <= 10^9
 * 0 <= d <= 10^9
 */

function maxPossibleScore(start: number[], d: number): number {
    start.sort((a, b) => a - b);  // Sort the array

    function check(score: number): boolean {
        let preX = -Infinity;  // Initial value for preX

        for (let s of start) {
            let x = preX + score;
            if (x > s + d) {
                return false;
            }
            preX = Math.max(x, s);
        }
        return true;
    }

    let left = 0;
    let right = (start[start.length - 1] + d - start[0]) + 1

    while (left + 1 < right) {
        const mid = Math.floor((left + right) / 2);
        if (check(mid)) {
            left = mid;
        } else {
            right = mid;
        }
    }
    return left;
}
